File: backend-springboot/src/main/java/com/monitor/controller/PerformanceController.java
package com.monitor.controller;

import com.monitor.service.PerformanceService;
import com.monitor.service.ElasticsearchService;
import com.monitor.util.DateUtil;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * 性能监控控制器
 */
@RestController
@RequestMapping("/performance")
@RequiredArgsConstructor
public class PerformanceController {
    
    private final PerformanceService performanceService;
    private final ElasticsearchService elasticsearchService;
    
    @Value("${default-project-id:001}")
    private String defaultProjectId;
    
    @PostMapping("/report")
    public ResponseEntity<Map<String, Object>> report(@RequestBody Map<String, Object> body) {
        return handleReport(body);
    }
    
    @GetMapping("/report")
    public ResponseEntity<Map<String, Object>> reportGet(@RequestParam String data) {
        try {
            String decoded = java.net.URLDecoder.decode(data, java.nio.charset.StandardCharsets.UTF_8);
            com.fasterxml.jackson.databind.ObjectMapper mapper = new com.fasterxml.jackson.databind.ObjectMapper();
            Map<String, Object> body = mapper.readValue(decoded, Map.class);
            return handleReport(body);
        } catch (Exception e) {
            System.err.println("Performance report GET failed: " + e.getMessage());
            return ResponseEntity.ok(Map.of("success", false, "message", "数据解析失败"));
        }
    }
    
    private ResponseEntity<Map<String, Object>> handleReport(Map<String, Object> body) {
        try {
            performanceService.report(
                defaultProjectId,
                (String) body.getOrDefault("url", ""),
                (String) body.get("userId"),
                (String) body.get("sessionId"),
                body.get("loadTime") != null ? ((Number) body.get("loadTime")).longValue() : null,
                body.get("domReady") != null ? ((Number) body.get("domReady")).longValue() : null,
                body.get("fcp") != null ? ((Number) body.get("fcp")).longValue() : null,
                body.get("lcp") != null ? ((Number) body.get("lcp")).longValue() : null,
                body.get("fid") != null ? ((Number) body.get("fid")).longValue() : null,
                body.get("cls") != null ? ((Number) body.get("cls")).doubleValue() : null,
                body.get("dns") != null ? ((Number) body.get("dns")).longValue() : null,
                body.get("tcp") != null ? ((Number) body.get("tcp")).longValue() : null,
                body.get("ttfb") != null ? ((Number) body.get("ttfb")).longValue() : null
            );
            
            // 异步写入 Elasticsearch（不阻塞主流程）
            Map<String, Object> logData = new HashMap<>();
            logData.put("projectId", defaultProjectId);
            logData.put("type", "performance");
            logData.put("userId", body.get("userId"));
            logData.put("sessionId", body.get("sessionId"));
            logData.put("url", body.getOrDefault("url", ""));
            logData.put("timestamp", body.getOrDefault("timestamp", new java.util.Date().toInstant().toString()));
            logData.put("loadTime", body.get("loadTime"));
            logData.put("domReady", body.get("domReady"));
            logData.put("fcp", body.get("fcp"));
            logData.put("lcp", body.get("lcp"));
            logData.put("fid", body.get("fid"));
            logData.put("cls", body.get("cls"));
            logData.put("dns", body.get("dns"));
            logData.put("tcp", body.get("tcp"));
            logData.put("ttfb", body.get("ttfb"));
            logData.put("message", "Performance metrics: loadTime=" + body.getOrDefault("loadTime", 0) + "ms");
            
            try {
                elasticsearchService.writeLog(logData);
            } catch (Exception e) {
                System.out.println("⚠️ Elasticsearch write failed (non-blocking): " + e.getMessage());
            }
            
            return ResponseEntity.ok(Map.of("success", true));
        } catch (Exception e) {
            System.err.println("Performance report failed: " + e.getMessage());
            return ResponseEntity.ok(Map.of("success", false, "message", "上报失败"));
        }
    }
    
    @GetMapping("/list")
    public ResponseEntity<Map<String, Object>> list(
            @RequestParam(required = false, defaultValue = "1") Integer page,
            @RequestParam(required = false, defaultValue = "20") Integer pageSize) {
        
        // TODO: 实现性能数据列表查询
        return ResponseEntity.ok(Map.of(
            "success", true,
            "data", List.of(),
            "total", 0,
            "page", page,
            "pageSize", pageSize
        ));
    }
    
    @GetMapping("/stats")
    public ResponseEntity<Map<String, Object>> stats() {
        Map<String, Object> stats = performanceService.getStats(defaultProjectId);
        return ResponseEntity.ok(Map.of("success", true, "data", stats));
    }
}

File: backend-midway/src/controller/performance.controller.ts
import { Controller, Get, Post, Inject, Body, Query, Config } from '@midwayjs/core';
import { ApiTags, ApiOperation, ApiQuery, ApiBody } from '@midwayjs/swagger';
import { PerformanceService } from '../service/performance.service';
import { ElasticsearchService } from '../service/elasticsearch.service';
import { formatDateTime } from '../utils/date.util';

@Controller('/performance')
@ApiTags(['performance'])
export class PerformanceController {
  @Inject()
  performanceService: PerformanceService;

  @Inject()
  elasticsearchService: ElasticsearchService;

  @Config('defaultProjectId')
  defaultProjectId: string;

  @Post('/report')
  @ApiOperation({ summary: '上报性能数据' })
  @ApiBody({ description: '性能指标数据' })
  async report(@Body() body: any) {
    return this.handleReport(body);
  }

  @Get('/report')
  @ApiOperation({ summary: '上报性能数据（GET方式，用于img上报）' })
  @ApiQuery({ name: 'data', required: true, description: 'JSON编码的性能数据' })
  async reportGet(@Query('data') dataStr: string) {
    try {
      const body = JSON.parse(decodeURIComponent(dataStr));
      return this.handleReport(body);
    } catch (error) {
      console.error('Performance report GET failed:', error);
      return { success: false, message: '数据解析失败' };
    }
  }

  private async handleReport(body: any) {
    try {
      await this.performanceService.report({
        projectId: this.defaultProjectId,
        url: body.url || '',
        userId: body.userId,
        sessionId: body.sessionId,
        loadTime: body.loadTime || 0,
        domReady: body.domReady,
        fcp: body.fcp,
        lcp: body.lcp,
        fid: body.fid,
        cls: body.cls,
        dns: body.dns,
        tcp: body.tcp,
        ttfb: body.ttfb,
      });

      // 异步写入 Elasticsearch（不阻塞主流程）
      this.elasticsearchService.writeLog({
        projectId: this.defaultProjectId,
        type: 'performance',
        userId: body.userId,
        sessionId: body.sessionId,
        url: body.url || '',
        timestamp: body.timestamp || new Date().toISOString(),
        loadTime: body.loadTime || 0,
        domReady: body.domReady,
        fcp: body.fcp,
        lcp: body.lcp,
        fid: body.fid,
        cls: body.cls,
        dns: body.dns,
        tcp: body.tcp,
        ttfb: body.ttfb,
        message: `Performance metrics: loadTime=${body.loadTime || 0}ms`,
      }).catch((err) => {
        // Elasticsearch 写入失败不影响主流程，只记录错误
        console.warn('⚠️ Elasticsearch write failed (non-blocking):', err.message);
      });

      return { success: true };
    } catch (error) {
      console.error('Performance report failed:', error);
      return { success: false, message: '上报失败' };
    }
  }

  @Get('/list')
  @ApiOperation({ summary: '获取性能数据列表' })
  @ApiQuery({ name: 'page', required: false, description: '页码' })
  @ApiQuery({ name: 'pageSize', required: false, description: '每页条数' })
  async list(@Query() query: any) {
    const { page, pageSize } = query;

    const result = await this.performanceService.getList({
      projectId: this.defaultProjectId,
      page: parseInt(page) || 1,
      pageSize: parseInt(pageSize) || 20,
    });

    return {
      success: true,
      data: result.data.map(item => ({
        ...item,
        timestamp: formatDateTime(item.timestamp),
      })),
      total: result.total,
      page: parseInt(page) || 1,
      pageSize: parseInt(pageSize) || 20,
    };
  }

  @Get('/stats')
  @ApiOperation({ summary: '获取性能统计' })
  async stats() {
    const stats = await this.performanceService.getStats(this.defaultProjectId);
    return { success: true, data: stats };
  }
}

